Implement explicit lockfiles integration

jaimergp opened the pull request:
Closes #21 

- [X] `conda list --explicit` extensions
- [x] `conda install | create` extensions for the above output
- [x] Tests
- [x] Docs

jaimergp commented:
This is how the output of `conda list --explicit` would work, btw:

```
# This file may be used to create an environment using:
# $ conda create --name <env> --file <this file>
# platform: osx-arm64
@EXPLICIT
https://conda.anaconda.org/conda-forge/osx-arm64/bzip2-1.0.8-h93a5062_5.conda
https://conda.anaconda.org/conda-forge/osx-arm64/c-ares-1.23.0-h93a5062_0.conda
https://conda.anaconda.org/conda-forge/osx-arm64/ca-certificates-2024.2.2-hf0a4a13_0.conda
https://conda.anaconda.org/conda-forge/osx-arm64/icu-73.2-hc8870d7_0.conda
https://conda.anaconda.org/conda-forge/osx-arm64/jpeg-9e-h1a8c8d9_3.conda
https://conda.anaconda.org/conda-forge/osx-arm64/libcxx-16.0.6-h4653b0c_0.conda
https://conda.anaconda.org/conda-forge/osx-arm64/libdeflate-1.17-h1a8c8d9_0.conda
https://conda.anaconda.org/conda-forge/osx-arm64/libedit-3.1.20191231-hc8eb9b7_2.tar.bz2
https://conda.anaconda.org/conda-forge/osx-arm64/libev-4.33-h93a5062_2.conda
https://conda.anaconda.org/conda-forge/osx-arm64/libffi-3.4.2-h3422bc3_5.tar.bz2
https://conda.anaconda.org/conda-forge/osx-arm64/libiconv-1.17-h0d3ecfb_1.conda
https://conda.anaconda.org/conda-forge/osx-arm64/libwebp-base-1.3.2-hb547adb_0.conda
https://conda.anaconda.org/conda-forge/osx-arm64/libzlib-1.2.13-h53f4e23_5.conda
https://conda.anaconda.org/conda-forge/osx-arm64/llvm-openmp-17.0.6-hcd81f8e_0.conda
https://conda.anaconda.org/conda-forge/osx-arm64/lzo-2.10-h642e427_1000.tar.bz2
# Following lines added by conda-pypi
# pypi: gitpython==3.1.32 --python-version 3.9.9 --implementation cp
# pypi: platformdirs==4.2.2 --python-version 3.9.9 --implementation cp
# pypi: sphinx==4.2.0 --python-version 3.9.9 --implementation cp
# pypi: python-dateutil==2.8.1 --python-version 3.9.9 --implementation cp
# pypi: pathspec==0.8.1 --python-version 3.9.9 --implementation cp
# pypi: charset-normalizer==2.0.7 --python-version 3.9.9 --implementation cp
# pypi: tornado==6.1 --python-version 3.9.9 --implementation cp --abi cp39 --platform macosx_11_0_arm64
# pypi: gitdb==4.0.10 --python-version 3.9.9 --implementation cp
# pypi: black==21.6b0 --python-version 3.9.9 --implementation cp
# pypi: regex==2021.4.4 --python-version 3.9.9 --implementation cp --abi cp39 --platform macosx_11_0_arm64
# pypi: tzdata==2023.3 --python-version 3.9.9 --implementation cp
# pypi: pillow==8.3.2 --python-version 3.9.9 --implementation cp --abi cp39 --platform macosx_11_0_arm64
```

We can't print the URL for PyPI wheels because it wasn't kept around 😬

jaimergp commented:
I'm not personally a fan of using the CLI flags here but these are the problems I faced trying to obtain something else:

- There's no record of the origin URL for an installed PyPI package (see https://github.com/pypa/pip/issues/10736). This might change with [PEP-710](https://peps.python.org/pep-0710/), but it's still a draft.
- The environment markers (`name=version; python_version='2.7' etc) do not limit which wheels will be obtained, but which lines are accepted for inclusion in the request list given the current Python interpreter.

I wish there was a way to encode the constraints in a single string field (like a URI or something), but right now this is the only way, I think.

jezdez commented on conda_pypi/main.py:
```diff
@@ -168,3 +194,138 @@ def ensure_target_env_has_externally_managed(command: str):
+                ignore = True
+                break
+            if path.name == "direct_url.json" and path.parent.suffix == ".dist-info":
+                data = json.loads(path.read_text())
```
this could raise OSError with permissions errors etc and should be handled more defensively

jezdez commented on conda_pypi/main.py:
```diff
@@ -168,3 +194,138 @@ def ensure_target_env_has_externally_managed(command: str):
+            lines.append(f"# pypi: {record.name}=={record.version}")
+            lines[-1] += f" --python-version {python_details['version']}"
+            lines[-1] += f" --implementation {python_details['implementation']}"
+            if wheel and (wheel_tag := wheel.get("Tag")):
+                for tag in wheel_tag:
+                    _, abi_tag, platform_tag = tag.split("-", 2)
+                    if abi_tag != "none" and abi_tag not in seen["abi"]:
+                        lines[-1] += f" --abi {abi_tag}"
+                        seen["abi"].add(abi_tag)
+                    if platform_tag != "any" and platform_tag not in seen["platform"]:
+                        lines[-1] += f" --platform {platform_tag}"
+                        seen["platform"].add(platform_tag)
```
Oh no, manually parsing tag info :( This must be available in packaging somewhere

jaimergp replied:
True. Using `packaging.tags.parse_tag` now. Thanks!

jezdez commented on conda_pypi/main.py:
```diff
@@ -168,3 +194,138 @@ def ensure_target_env_has_externally_managed(command: str):
+    return lines
+
+
+def dry_run_pip_json(args: Iterable[str], force_reinstall: bool = False) -> dict[str, Any]:
```
needs docs

jezdez commented on conda_pypi/main.py:
```diff
@@ -168,3 +194,138 @@ def ensure_target_env_has_externally_managed(command: str):
+    # pip can output to stdout via `--report -` (dash), but this
+    # creates issues on Windows due to undecodable characters on some
+    # project descriptions (e.g. charset-normalizer, amusingly), which
+    # makes pip crash internally. Probably a bug on their end.
```
let's file a bug upstream

jezdez commented on conda_pypi/main.py:
```diff
@@ -168,3 +194,138 @@ def ensure_target_env_has_externally_managed(command: str):
+        "install",
+        "--dry-run",
+        "--ignore-installed",
+        *(("--force-reinstall",) if force_reinstall else ()),
```
what's with the unpacking?

```suggestion
        "--force-reinstall" if force_reinstall else "",
```

jaimergp replied:
I don't want to pass an empty string which subprocess will quote as a positional arg, won't it?

jaimergp replied:
Changed to conditional extensions here in general.

jezdez commented on conda_pypi/main.py:
```diff
@@ -168,3 +194,138 @@ def ensure_target_env_has_externally_managed(command: str):
+        "--report",
+        json_output.name,
+        "--target",
+        json_output.name + ".dir",
```
why .dir?

jezdez commented on conda_pypi/main.py:
```diff
@@ -168,3 +194,138 @@ def ensure_target_env_has_externally_managed(command: str):
+        # contain weird unicode stuff (as part of the project
+        # description or README).
+        report = json.loads(f.read())
+    os.unlink(json_output.name)
```
this should be done in a big try/finally block to make sure this file is deleted no matter what.

Also the `*.dir` directory created by the --target option above needs cleaning up

jaimergp replied:
Done.

> Also the *.dir directory created by the --target option above needs cleaning up

It's never actually created, so it fails to clean up. I didn't add it.

jezdez commented on conda_pypi/main.py:
```diff
@@ -168,3 +194,138 @@ def ensure_target_env_has_externally_managed(command: str):
+
+
+def compute_record_sum(record_path, algo):
+    record = Path(record_path).read_text()
```
OSError again

jezdez commented on conda_pypi/main.py:
```diff
@@ -168,3 +194,138 @@ def ensure_target_env_has_externally_managed(command: str):
+            # we only want to check the metadata and wheel parts of dist-info; everything else
+            # is not deterministic or useful
+            continue
+        if path.parts[0] == ".." and any(part in path.parts for part in ("bin", "lib", "Scripts")):
```
```suggestion
        elif path.parts[0] == ".." and any(part in path.parts for part in ("bin", "lib", "Scripts")):
```

jezdez reviewed (commented):
A few notes:
- I'm not sure how I feel you expanding this format without a spec and a CEP since that's definitely a core functionality in conda
- There has to be other ways than implementing custom argv parsing and custom Python packaging standards parsing (big maintenance risk based on my experience)
- Inline docs and high-level reasoning for some of the design decisions are sorely needed before merging this

jezdez reviewed (commented):
A few notes:
- I'm not sure how I feel you expanding this format without a spec and a CEP since that's definitely a core functionality in conda
- There has to be other ways than implementing custom argv parsing and custom Python packaging standards parsing (big maintenance risk based on my experience)
- Inline docs and high-level reasoning for some of the design decisions are sorely needed before merging this

jaimergp commented:
Thanks for the thorough review!

> I'm not sure how I feel you expanding this format without a spec and a CEP since that's definitely a core functionality in conda

I do want the CEP. For the original format too. There's no standard to refer to other than the implementation. It should be straightforward.

> There has to be other ways than implementing custom argv parsing and custom Python packaging standards parsing (big maintenance risk based on my experience)

`conda` has [its own parsers](https://github.com/conda/conda/blob/main/conda/common/pkg_formats/python.py) but I only learnt that after writing this prototype. I'm sure we can find some parsers somewhere in `packaging` and friends too. Right now I'm mostly exploring how this idea would shape up, but I agree we need to polish this custom code better. Even the structure of the code could use some improvement.

> Inline docs and high-level reasoning for some of the design decisions are sorely needed before merging this

Yep, just wanted to get some external confirmation that this idea is worth pursuing before writing details on something that might not fly. I take that this review is also a 👍 to proceed further?

jaimergp commented on conda_pypi/cli/install.py:
```diff
@@ -0,0 +1,111 @@
+if TYPE_CHECKING:
+    from typing import Iterable
+
+log = getLogger(f"conda.{__name__}")
```
I've been doing this in all my plugins to reuse `conda`'s configuration for consistency. CLS does this too, for example.

jaimergp commented on conda_pypi/cli/list.py:
```diff
@@ -0,0 +1,22 @@
+    if command != "list":
+        return
+    cmd_line = context.raw_data.get("cmd_line", {})
+    if "--explicit" not in sys.argv and not cmd_line.get("explicit"):
+        return
+    if "--no-pip" in sys.argv or not cmd_line.get("pip"):
+        return
```
I've seen this kind of suggestion before but never understood why. Isn't the `return` above enough to guarantee the lack of execution? To me, using `if` explicitly emphasizes the early return, but I can also see how using `elif` better reveals the conditional relationship between the statements.

jaimergp commented on conda_pypi/cli/list.py:
```diff
@@ -0,0 +1,22 @@
+    if command != "list":
+        return
+    cmd_line = context.raw_data.get("cmd_line", {})
+    if "--explicit" not in sys.argv and not cmd_line.get("explicit"):
```
I wish we had a good API for the parsed args here. I want to use `context.raw_data` explicitly but I fear this is a data structure API we are not watching closely. See https://github.com/conda/conda/issues/13924.

jaimergp replied:
Changed so we only use `context.raw_data` now.

jaimergp commented on conda_pypi/main.py:
```diff
@@ -168,3 +194,138 @@ def ensure_target_env_has_externally_managed(command: str):
+    pd = PrefixData(str(prefix), pip_interop_enabled=True)
+    pd.load()
+    lines = []
+    python_record = list(pd.query("python"))
+    assert len(python_record) == 1
+    python_record = python_record[0]
+    python_details = {"version": ".".join(python_record.version.split(".")[:3])}
+    if "pypy" in python_record.build:
+        python_details["implementation"] = "pp"
+    else:
+        python_details["implementation"] = "cp"
```
I'm not aware of any other impls in conda-forge, or actually any system to query the implementation robustly across channels without executing code. I'll need to look further into this.

We do have `python_abi` and so on, but that's conda-forge specific, isn't it? Probably worth moving to its own utility function. I'll refactor this function a bit to accommodate the other changes too.

jaimergp replied:
I extended this check so it explicitly checks for `cpython` too, and falls back to None if no matches were found.

jaimergp commented on conda_pypi/cli/install.py:
```diff
@@ -0,0 +1,111 @@
+                    if (dist_info / "RECORD").is_file():
+                        found_hash = compute_record_sum(dist_info / "RECORD", algo)
+                        if expected_hash != found_hash:
+                            log.warning(
```
I think conda has a couple options to disable checks. I think we can honor those here. And actually should. Let me add those.

jaimergp replied:
Done.

jaimergp commented on conda_pypi/main.py:
```diff
@@ -168,3 +194,138 @@ def ensure_target_env_has_externally_managed(command: str):
+                    if key == "Tag":
+                        wheel.setdefault(key, []).append(value.strip())
+                    else:
+                        wheel[key] = value.strip()
```
I dug in the docs and it uses RFC email headers, so `email.parser.HeaderParser` from stdlib can be used. It requires a bit of post-processing, quite similar to what we see in [`conda.common.pkg_formats.python`](https://github.com/conda/conda/blob/a261ade3d0a276a214bff314e54c9a9dbb8c282e/conda/common/pkg_formats/python.py#L561). That module would better be here in this project now. That whole thing is mostly there to support `conda list` listing `pypi` dependencies after all. We can refactor it out and finally remove the `pip_interop_enabled` thingy.

jaimergp commented on conda_pypi/cli/install.py:
```diff
@@ -0,0 +1,111 @@
+        if arg == "--file":
+            pypi_lines += [
+                line[pypi_prefix_len:]
+                for line in Path(argv[i + 1]).read_text().splitlines()
```
I added a safety check just in case now.

jaimergp commented on conda_pypi/cli/install.py:
```diff
@@ -0,0 +1,111 @@
+
+def _prepare_pypi_transaction(pypi_lines) -> dict[str, dict[str, str]]:
+    pkgs = {}
+    for args in pypi_lines:
+        args = shlex.split(args)
+        record_hash = None
+        if "--" in args:
```
I changed this part and provided an argparser instance.

jaimergp commented:
> I do want the CEP. For the original format too.

There you go: https://github.com/conda/ceps/pull/79

jaimergp commented:
Hey @jezdez, I think I addressed most of your comments now. The PR is getting a bit too big so we can address further changes in new PRs. WDYT? After this PR I feel comfortable moving things to `conda-incubator`.

jezdez commented:
@jaimergp thank you for doing the rounds on the standardization, this'll be a net benefit!
